src/expr/src/table_function/mod.rs

```diff
@@ -62,3 +73,184 @@ pub fn build_from_prost(prost: &TableFunctionProst) -> Result<BoxedTableFunction
+        fn return_type(&self) -> DataType {
+            self.expr.return_type()
+        }
+
+        fn eval(&self, input: &DataChunk) -> Result<Vec<ArrayRef>> {
+            let array = self.expr.eval(input)?;
+
+            let mut res = vec![];
+            for datum_ref in array.iter() {
+                let mut builder = self.return_type().create_array_builder(self.n);
+                for _ in 0..self.n {
+                    builder.append_datum_ref(datum_ref)?;
+                }
+                let array = builder.finish()?;
+                res.push(Arc::new(array));
+            }
+
+            Ok(res)
+        }
+    }
+
+    Mock { expr, n }.boxed()
+}
+
+#[derive(Debug)]
+pub enum ProjectSetSelectItem {
+    TableFunction(BoxedTableFunction),
+    Expr(BoxedExpression),
+}
+
+impl From<BoxedTableFunction> for ProjectSetSelectItem {
+    fn from(table_function: BoxedTableFunction) -> Self {
+        ProjectSetSelectItem::TableFunction(table_function)
+    }
+}
+
+impl From<BoxedExpression> for ProjectSetSelectItem {
+    fn from(expr: BoxedExpression) -> Self {
+        ProjectSetSelectItem::Expr(expr)
+    }
+}
+
+/// A column corresponds to the range `0..len`
+fn index_array_column(len: usize) -> Column {
+    let mut builder = I64ArrayBuilder::new(len);
+    for value in 0..len {
+        builder.append(Some(value as i64)).unwrap();
+    }
+    let array = builder.finish().unwrap();
+    Column::new(Arc::new(array.into()))
+}
+
+impl ProjectSetSelectItem {
+    pub fn from_prost(prost: &SelectItemProst) -> Result<Self> {
+        match prost.select_item.as_ref().unwrap() {
+            Expr(expr) => expr_build_from_prost(expr).map(Into::into),
+            TableFunction(tf) => build_from_prost(tf).map(Into::into),
+        }
+    }
+
+    pub fn return_type(&self) -> DataType {
+        match self {
+            ProjectSetSelectItem::TableFunction(tf) => tf.return_type(),
+            ProjectSetSelectItem::Expr(expr) => expr.return_type(),
+        }
+    }
+
+    fn eval(&self, input: &DataChunk) -> Result<ProjectSetSelectItemResult> {
+        match self {
+            ProjectSetSelectItem::TableFunction(tf) => tf
+                .eval(input)
+                .map(ProjectSetSelectItemResult::TableFunction),
+            ProjectSetSelectItem::Expr(expr) => {
+                expr.eval(input).map(ProjectSetSelectItemResult::Expr)
+            }
+        }
+    }
+
+    /// First column will be `projected_row_id`, which represents the index in the output table
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    pub async fn execute(
+        select_list: Arc<Vec<Self>>,
+        data_types: Vec<DataType>,
+        input: &DataChunk,
+    ) {
+        assert!(!select_list.is_empty());
+
+        let results: Vec<_> = select_list
+            .iter()
+            .map(|select_item| select_item.eval(input))
+            .try_collect()?;
+
+        let mut iters = results.iter().map(|r| r.iter()).collect_vec();
```

Comment:
From my perspective, it's easier to simply access the array by index 🤣, and we won't need to have `ProjectSetSelectItemResultIter`. Anyway, the current solution is also okay to me.